rclpy/rclpy/qos_event.py

```diff
@@ -0,0 +1,226 @@
+    """
+
+    RCL_SUBSCRIPTION_REQUESTED_DEADLINE_MISSED = 0
+    RCL_SUBSCRIPTION_LIVELINESS_CHANGED = 1
+
+
+"""
+Payload type for Subscription Deadline callback.
+
+Mirrors rmw_requested_deadline_missed_status_t from rmw/types.h
+"""
+QoSRequestedDeadlineMissedInfo = NamedTuple(
+    'QoSRequestedDeadlineMissedInfo', [
+        ('total_count', 'int'),
+        ('total_count_change', 'int'),
+    ])
+
+"""
+Payload type for Subscription Liveliness callback.
+
+Mirrors rmw_liveliness_changed_status_t from rmw/types.h
+"""
+QoSLivelinessChangedInfo = NamedTuple(
+    'QoSLivelinessChangedInfo', [
+        ('alive_count', 'int'),
+        ('not_alive_count', 'int'),
+        ('alive_count_change', 'int'),
+        ('not_alive_count_change', 'int'),
+    ])
+
+"""
+Payload type for Publisher Deadline callback.
+
+Mirrors rmw_offered_deadline_missed_status_t from rmw/types.h
+"""
+QoSOfferedDeadlineMissedInfo = NamedTuple(
+    'QoSOfferedDeadlineMissedInfo', [
+        ('total_count', 'int'),
+        ('total_count_change', 'int'),
+    ])
+
+"""
+Payload type for Publisher Liveliness callback.
+
+Mirrors rmw_liveliness_lost_status_t from rmw/types.h
+"""
+QoSLivelinessLostInfo = NamedTuple(
+    'QoSLivelinessLostInfo', [
+        ('total_count', 'int'),
+        ('total_count_change', 'int'),
+    ])
+
+
+class QoSEventHandler(Waitable):
+    """Waitable type to handle QoS events."""
+
+    def __init__(
+        self,
+        *,
+        callback_group,
+        callback,
+        event_type,
+        parent_handle,
+    ):
+        # Waitable init adds self to callback_group
+        super().__init__(callback_group)
+        self.event_type = event_type
+        self.callback = callback
+
+        self._parent_handle = parent_handle
+        self._event_handle = _rclpy.rclpy_create_event(event_type, parent_handle)
+        self._ready_to_take_data = False
+        self._event_index = None
+
+    # Start Waitable API
+    def is_ready(self, wait_set):
+        """Return True if entities are ready in the wait set."""
+        if self._event_index is None:
+            return False
+        if _rclpy.rclpy_wait_set_is_ready('event', wait_set, self._event_index):
+            self._ready_to_take_data = True
+        return self._ready_to_take_data
+
+    def take_data(self):
+        """Take stuff from lower level so the wait set doesn't immediately wake again."""
+        if self._ready_to_take_data:
+            self._ready_to_take_data = False
+            return _rclpy.rclpy_take_event(
+                self._event_handle, self._parent_handle, self.event_type)
+        return None
+
+    async def execute(self, taken_data):
+        """Execute work after data has been taken from a ready wait set."""
+        if not taken_data:
+            return
+        self.callback(taken_data)
+
+    def get_num_entities(self):
+        """Return number of each type of entity used."""
+        return NumberOfEntities(num_events=1)
```

Comment:
Since there is just one entity, this might fit better as an `Event` class instead of using the `Waitable` API. Though, that makes adding it to the executor harder so I'm fine with this as is.

Reply:
Noted - maybe we will change it in a future revision, for now I will leave it as a Waitable